receiver/awscloudwatchreceiver/logs.go

```diff
@@ -0,0 +1,335 @@
+}
+
+type namesPollConfig struct {
+	logGroup    string
+	streamNames []*string
+}
+
+func (npc *namesPollConfig) request(limit int64, st, et *time.Time) *cloudwatchlogs.FilterLogEventsInput {
+	base := &cloudwatchlogs.FilterLogEventsInput{
+		LogGroupName: &npc.logGroup,
+		StartTime:    aws.Int64(st.UnixMilli()),
+		EndTime:      aws.Int64(et.UnixMilli()),
+		Limit:        aws.Int64(limit),
+	}
+	if len(npc.streamNames) > 0 {
+		base.LogStreamNames = npc.streamNames
+	}
+	return base
+}
+
+type prefixPollConfig struct {
+	logGroup string
+	prefix   *string
+}
+
+func (ppc *prefixPollConfig) request(limit int64, st, et *time.Time) *cloudwatchlogs.FilterLogEventsInput {
+	return &cloudwatchlogs.FilterLogEventsInput{
+		LogGroupName:        &ppc.logGroup,
+		StartTime:           aws.Int64(st.UnixMilli()),
+		EndTime:             aws.Int64(et.UnixMilli()),
+		Limit:               aws.Int64(limit),
+		LogStreamNamePrefix: ppc.prefix,
+	}
+}
+
+type pollConfig interface {
+	request(limit int64, startTime, endTime *time.Time) *cloudwatchlogs.FilterLogEventsInput
+}
+
+func newLogsReceiver(cfg *Config, logger *zap.Logger, consumer consumer.Logs) *logsReceiver {
+	namedPolls := []namesPollConfig{}
+	prefixedPolls := []prefixPollConfig{}
+	for logGroupName, sc := range cfg.Logs.Groups.NamedConfigs {
+		for _, prefix := range sc.Prefixes {
+			prefixedPolls = append(prefixedPolls, prefixPollConfig{
+				logGroup: logGroupName,
+				prefix:   prefix,
+			})
+		}
+
+		namedPolls = append(namedPolls, namesPollConfig{
+			logGroup:    logGroupName,
+			streamNames: sc.Names,
+		})
+	}
+
+	// safeguard from using both
+	autodiscover := cfg.Logs.Groups.AutodiscoverConfig
+	if len(cfg.Logs.Groups.NamedConfigs) > 0 {
+		autodiscover = nil
+	}
+
+	return &logsReceiver{
+		region:              cfg.Region,
+		profile:             cfg.Profile,
+		consumer:            consumer,
+		maxEventsPerRequest: cfg.Logs.MaxEventsPerRequest,
+		imdsEndpoint:        cfg.IMDSEndpoint,
+		autodiscover:        autodiscover,
+		pollInterval:        cfg.Logs.PollInterval,
+		namedPolls:          namedPolls,
+		prefixedPolls:       prefixedPolls,
+		logger:              logger,
+		wg:                  &sync.WaitGroup{},
+		doneChan:            make(chan bool),
+	}
+}
+
+func (l *logsReceiver) Start(ctx context.Context, host component.Host) error {
+	l.logger.Debug("starting to poll for Cloudwatch logs")
+
+	if l.autodiscover != nil {
+		namedPolls, prefixedPolls, err := l.discoverGroups(ctx, l.autodiscover)
+		if err != nil {
+			return fmt.Errorf("unable to auto discover log groups using prefix  %s: %w", l.autodiscover.Prefix, err)
+		}
+		l.namedPolls = namedPolls
+		l.prefixedPolls = prefixedPolls
+	}
+
+	for i := range l.namedPolls {
+		np := l.namedPolls[i]
+		l.wg.Add(1)
+		go l.poll(ctx, np.logGroup, &np)
+	}
+
+	for i := range l.prefixedPolls {
+		pp := l.prefixedPolls[i]
+		l.wg.Add(1)
+		go l.poll(ctx, pp.logGroup, &pp)
+	}
```

Comment:
Is it really beneficial to parallelize these? Are you sure this won't be problematic at scale? Why not just process each group in series?

Reply:
I unparallelized it 👍 I think this should be less of a concern now